File: memento/golang/dish.go
package main

// Dish represents the Originator with an internal state
type Dish struct {
	Name string
}

// DishMemento is the concrete Memento storing the Dish state
type DishMemento struct {
	name string
}

func (d *Dish) Save() *DishMemento {
	return &DishMemento{name: d.Name}
}

func (d *Dish) Restore(m *DishMemento) {
	d.Name = m.name
}

File: memento/rust/memento/src/dish.rs
#[derive(Clone, Debug)]
pub struct Dish {
    pub name: String,
    pub stage: String, // e.g., "Prepped", "Cooking", "Garnished"
}

// The Memento for Dish
#[derive(Clone)]
pub struct DishMemento {
    state: String,
}

impl Dish {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            stage: "Raw".to_string(),
        }
    }

    pub fn save(&self) -> DishMemento {
        DishMemento { state: self.stage.clone() }
    }

    pub fn restore(&mut self, memento: &DishMemento) {
        self.stage = memento.state.clone();
    }
}
